contribute/design/1.x/replica_scaleup/dataplane/20190910-replica-scaleup.md

```diff
@@ -0,0 +1,278 @@
+- operator that detects the need to increase ReplicationFactor
+- Allow adding replicas to replace particular replica (volume distribution case)
+
+## Proposal
+
+### User Stories
+
+#### Add new replica
+As an OpenEBS user, I should be able to add replicas to my volume.
+
+#### Replace non-existing replica
+As an OpenEBS user, I should be able to replace non-existing replica with new
+replica.
+
+### Implementation Details/Notes/Constraints
+
+Currently, cstor-istgt reads replication related configuration, i.e.,
+ReplicationFactor (RF) and ConsistencyFactor (CF), from istgt.conf file.
+cstor-volume-mgmt is the container that fills the RF and CF details into
+istgt.conf file by reading from CStorVolume CR.
+
+Another property that cstor-istgt uses is 'quorum'. This property is set at
+replica, and, replica sends this to istgt during handshake phase.
+This property can take values of 'on' and 'off'.
+'on' means that data written to this replica is available for reads.
+'off' means that data written to this replica is lost.
+cstor-pool-mgmt is the container that sets quorum property at replica.
+It sets property as 'on' if replica is created first time related to that CVR.
+Otherwise, it sets to 'off'.
+
+#### Current Implementation -- Config
+- cstor-volume-mgmt reads configuration from CStorVolume CR.
+  - spec.replicationFactor
+  - spec.consistencyFactor
+- cstor-volume-mgmt updates the above info into istgt.conf file as
+  - ReplicationFactor
+  - ConsistencyFactor
+- Based on the status of CVR and availability of replica, cstor-pool-mgmt
+sets quorum property while creating replica.
+
+#### Current Implementation
+Above mentioned properties are used as below at istgt:
+- Allow replicas (upto max of 5) to connect to target until RF number of
+healthy replicas are available.
+- Atleast CF in-quorum replicas are required to perform rebuilding and start IOs
+- Rebuilding will be triggered and replicas become healthy. As RF number of
+replicas are healthy, other connected replicas, if any, gets disconnected.
+
+#### Shortcomings With Current Implementation
+- Due to misconfiguration from user, if old replica connects back instead of
+the replaced one, there are chances of serving wrong data, and can lead to
+data corruption.
+- There is no data-consistent way of increasing RF and CF of CStorVolume CR.
+
+### Proposed Implementation
+Increase in replication factor will be provided in declarative way.
+For replicas identification that can connect to target, replicas information
+as well need to be stored in CStorVolume CR and istgt.conf.
+
+New fields will be added to CStorVolume CR and istgt.conf.
+In CStorVolume CR, spec.DesiredReplicationFactor will be added to help in
+adding replicas, and, status.ReplicaList to store list of known replicas.
+In istgt.conf, DesiredReplicationFactor and ReplicaList will be added.
+
+
+#### Adding new replica
+- User will create proper CVR with other than Empty status.phase
```

Comment:
Should the status be always "Init" ?

Reply:
CVR controller seems to be using Empty, i.e., "" status.Phase to find whether CVR is newly created or not.
So, if status.Phase of CVR is Empty, CVR controller creates the replica (as zvol is not created yet) with quorum ON.
If status.Phase of CVR is other than Empty, CVR controller creates the replica if zvol is already not created (this happens when pool got recreated) with quorum Off.
In this case of adding new replica, we want replica to be added with quorum Off. So, status of CVR CR should be other than Empty status.Phase.
I will update OEP with above details, and also, particular value for the status rather than saying 'other than Empty'